myia/public_api.py

```diff
@@ -523,10 +527,16 @@ def norm(inp, p=None, dim=None):
     if p is None:
         p = 2
 
-    # Convert p to input type.
-    p = P.scalar_to_array(P.scalar_cast(p, inp.dtype), APT)
-    # Create a value 1 with input type.
-    one = P.scalar_to_array(P.scalar_cast(1, inp.dtype), APT)
+    if P.hastype(inp, APT):
+        # Convert p to input type.
+        p = P.scalar_to_array(P.scalar_cast(p, inp.dtype), APT)
+        # Create a value 1 with input type.
+        one = P.scalar_to_array(P.scalar_cast(1, inp.dtype), APT)
+    else:
+        # Convert p to input type.
+        p = P.scalar_to_array(P.scalar_cast(p, inp.dtype), AA)
+        # Create a value 1 with input type.
+        one = P.scalar_to_array(P.scalar_cast(1, inp.dtype), AA)
```

Comment:
I think this should work:

```suggestion
    # Convert p to input type.
    p = P.scalar_to_array(P.scalar_cast(p, inp.dtype), operations.typeof(inp))
    # Create a value 1 with input type.
    one = P.scalar_to_array(P.scalar_cast(1, inp.dtype), operations.typeof(inp))
```